src/libAtomVM/ets.c

```diff
@@ -286,29 +280,46 @@ static EtsErrorCode ets_table_insert(struct EtsTable *ets_table, term entry, Con
 
 static EtsErrorCode ets_table_insert_list(struct EtsTable *ets_table, term list, Context *ctx)
 {
+    if (ets_table->access_type != EtsAccessPublic && ets_table->owner_process_id != ctx->process_id) {
+        return EtsPermissionDenied;
+    }
     term iter = list;
+    size_t size = 0;
 
     while (term_is_nonempty_list(iter)) {
         term tuple = term_get_list_head(iter);
         iter = term_get_list_tail(iter);
         if (!term_is_tuple(tuple) || (size_t) term_get_tuple_arity(tuple) < (ets_table->keypos + 1)) {
             return EtsBadEntry;
         }
+        ++size;
     }
     if (!term_is_nil(iter)) {
         return EtsBadEntry;
     }
 
+    struct HNode **hnode_list = malloc(size * sizeof(struct HNode *));
+    if (IS_NULL_PTR(hnode_list)) {
+        return EtsAllocationFailure;
+    }
+
+    int cur = 0;
     while (term_is_nonempty_list(list)) {
```

Comment:
Can't we do it in the previous loop instead of iterating twice?

Reply:
Note: we don't know length of the list.

Reply:
no, not in an easy way: it is so hard to undo the operation inside the second loop. First loop checks validity, second applies it.